### ISSUE TITLE
IPP Everywhere: lpadmin crashes when discovering HP Printer

### ISSUE THREAD

**Issue by stephanwib** [apple/cups#4893]
Printer: HP LaserJet 500 color m551
Problem:
# lpadmin -p hp-e -v ipp://10.70.235.32/ -m everywhere
Memory fault (core dumped)

Analysis:
The issue is that the printer reports the following properites:

<key>media-col-default</key>
<dict>
<key>media-size</key>
<dict>
<key>x-dimension</key>
<integer>0</integer>
<key>y-dimension</key>
<integer>0</integer>
</dict>

This leads to a problem in the following code path in cups/ppd-cache.c:

--------------------------------------------------------------
  if ((defattr = ippFindAttribute(response, "media-col-default",
IPP_TAG_BEGIN_COLLECTION)) != NULL)
  {
    if ((attr = ippFindAttribute(ippGetCollection(defattr, 0),
"media-size", IPP_TAG_BEGIN_COLLECTION)) != NULL)
    {
      media_size = ippGetCollection(attr, 0);
      x_dim      = ippFindAttribute(media_size, "x-dimension", IPP_TAG_INTEGER);
      y_dim      = ippFindAttribute(media_size, "y-dimension", IPP_TAG_INTEGER);

      if (x_dim && y_dim)
      {
        pwg = pwgMediaForSize(ippGetInteger(x_dim, 0), ippGetInteger(y_dim, 0));
        strlcpy(ppdname, pwg->ppd, sizeof(ppdname));
      }
      else
        strlcpy(ppdname, "Unknown", sizeof(ppdname));
    }
-----------------------------------------------------------------------

The values of the dimensions are not checked and pwgMediaForSize()
returns a NULL pointer which is then accessed.

-------------------------------------------------------------------
(gdb) b pwgMediaForSize
Function "pwgMediaForSize" not defined.
Make breakpoint pending on future shared library load? (y or [n]) y
Breakpoint 1 (pwgMediaForSize) pending.
(gdb) r
Starting program: /opt/cups-2.2.1/sbin/lpadmin -p hp-e -v
ipp://10.70.235.32/ -m everywhere
Missing separate debuginfos, use: zypper install
glibc-debuginfo-2.19-31.9.x86_64
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib64/libthread_db.so.1".

Breakpoint 1, pwgMediaForSize (width=0, length=0) at pwg-media.c:911
911       return (_pwgMediaNearSize(width, length, 176));
(gdb) bt
#0  pwgMediaForSize (width=0, length=0) at pwg-media.c:911
#1  0x00007ffff7b96522 in _ppdCreateFromIPP (buffer=0x7fffffffa2f0
"/tmp/01cbc57fd2a31", bufsize=16384, response=0x55555576af60) at
ppd-cache.c:3151
#2  0x000055555555887c in get_printer_ppd (uri=0x55555575f0a4
"ipp://10.70.235.32/", buffer=0x7fffffffa2f0 "/tmp/01cbc57fd2a31",
bufsize=16384) at lpadmin.c:1204
#3  0x0000555555557345 in main (argc=7, argv=0x7fffffffe3f8) at lpadmin.c:596
(gdb) finish
Run till exit from #0  pwgMediaForSize (width=0, length=0) at pwg-media.c:911
0x00007ffff7b96522 in _ppdCreateFromIPP (buffer=0x7fffffffa2f0
"/tmp/01cbc57fd2a31", bufsize=16384, response=0x55555576af60) at
ppd-cache.c:3151
3151            pwg = pwgMediaForSize(ippGetInteger(x_dim, 0),
ippGetInteger(y_dim, 0));
Value returned is $1 = (pwg_media_t *) 0x0
(gdb) quit
----------------------------------------------------------

A patch is attached.
[patch-ppdcache.txt](https://github.com/apple/cups/files/513672/patch-ppdcache.txt)

**Comment by michaelrsweet** [comment]
Also reported on the cups-devel mailing list:

```
Hi!

I found another issue regarding IPP discovery and input trays. These
are my printers:

# ipptool ipp://10.70.235.31/ getattr.test
ipp-versions-supported printer-name media-source-supported
---------------------- ------------ -------------------------------------
1.0,1.1,2.0            Lexmark C748 auto,tray-1,alternate,manual,envelope

# ipptool ipp://10.70.235.32/ getattr.test
ipp-versions-supported printer-name media-source-supported
---------------------- ------------ ----------------------
1.0,1.1,2.0            NPIB80216    auto,tray-2,tray-1
(HP LaserJet 500 color m551)


I found that trays named tray-* were not included in the resulting PPD
and thus not selectable for the printer object. I suspect that is
because the tray descriptions are wrong in ppd-cache.c.

   static const char * const sources[][2] =
   ...
     { "Tray1", "Tray 1" },
     { "Tray2", "Tray 2" },
     { "Tray3", "Tray 3" },

I changed these as below, according to
https://ftp.pwg.org/pub/pwg/candidates/cs-ippjobprinterext3v10-20120727-5100.13.pdf
Table 14:

     { "Tray-1", "Tray 1" },
     { "Tray-2", "Tray 2" },
     { "Tray-3", "Tray 3" },
     ...

The trays now appear in the PPD and are also selectable:

*OrderDependency: 10 AnySetup *InputSlot
*DefaultInputSlot: Auto
*InputSlot Auto/Automatic: "<</MediaPosition 0>>setpagedevice"
*InputSlot Tray-2/Tray 2: "<</MediaPosition 21>>setpagedevice"
*InputSlot Tray-1/Tray 1: "<</MediaPosition 20>>setpagedevice"
*CloseUI: *InputSlot

However, when a print job is submitted to the printer, the
media-source job attribute is never sent, leaving tray selection not
working. The option is passed to the IPP backend, but is omitted by it
for some reason.

D [07/Oct/2016:07:01:42 +0000] [Job 297] argv[5]="colormodel=rgb
Duplex=none finishings=3 inputslot=tray-1 number-up=1 PageSiz
e=a4 job-uuid=urn:uuid:ecf5c718-230b-38ac-7e1e-119f1a5370f2
job-originating-host-name=localhost date-time-at-creation= date-ti
me-at-processing= time-at-creation=1475823702
time-at-processing=1475823702 document-name-supplied=testfile.pdf"

-Am I correct as for the tray descriptions in ppd-cache.c?

-What is the reason for the IPP backend not sending the media-source
job attribute?


Regards,

Stephan
```

**Comment by michaelrsweet** [comment]
And a followup:

```
I now understand how it needs to be. The format of the input slots
needs to look like this in the PPD:

*InputSlot Tray2/Tray2: "<</MediaPosition 21>>setpagedevice"
*InputSlot Tray1/Tray1: "<</MediaPosition 20>>setpagedevice"

The string (e.g. Tray2) is automatically lowered and a dash is
inserted which leads to a "tray-2" for the media-source attribute. I
don´t yet understand how these PPD to IPP conversion routines work in
detail.

I was able to get the tray discovery and selection working
out-of-the-box with the following hacks to ppd-cache.c


static const char * const sources[][2] =
...
     { "Tray-1", "Tray1" },
     { "Tray-2", "Tray2" },
     { "Tray-3", "Tray3" },

-         cupsFilePrintf(fp, "*InputSlot %s/%s: \"<</MediaPosition
%d>>setpagedevice\"\n", ppdname, sources[j][1], j);
+         cupsFilePrintf(fp, "*InputSlot %s/%s: \"<</MediaPosition
%d>>setpagedevice\"\n", sources[j][1], sources[j][1], j);

I have no idea how to fix this in the sense of the whole architecture
of the cups source.

Regards,

Stephan
```

**Comment by michaelrsweet** [comment]
[master 2a8afc2] Fixed some issues with IPP Everywhere printer support (Issue #4893)
